website/blog/2020-06-16-alluxio-datacaching.md

```diff
@@ -0,0 +1,243 @@
+
+Alluxio file system is an open-source data orchestration system that is often used as a distributed caching service to Presto. 
+To achieve sub-second query latencies in our architecture, we want to further reduce the communication overhead between Presto and Alluxio. 
+As a result, core teams from Alluxio and Presto collaborated to carve out a single-node, embedded cache library from Alluxio service.
+<br><br>
+In particular, a Presto worker queries this Alluxio local cache inside the same JVM through a standard HDFS interface. 
+On a cache hit, Alluxio local cache directly reads data from the local disk and returns cache result to the Presto threads; 
+otherwise, it retrieves data from the remote data source, and caches the data on the local disk for followup queries. 
+This cache is completely transparent to Presto. In case the cache runs into issues (e.g., local disk failures), the Presto reads 
+fallback to the remote data source. This workflow is shown as the figure below.
+
+<br><br>
+![](/img/blog/2020-06-16-alluxio-datacaching/presto_alluxio_caching.jpg)
+<br><br>
+
+## Cache internals and configuration
+Our Alluxio data cache is a library residing in the Presto worker. It provides an HDFS-compatible interface “AlluxioCachingFileSystem” as the 
+main interface to Presto workers for all data access operations. 
+<br>
+There are some design choices under the hood:
+
+### Basic Caching Unit
+Both Alluxio experience and earlier experiments from the Facebook team suggested that reading, writing and evicting data in a fixed block size is 
+most efficient. Though in the Alluxio system, the default caching block size, 64MB, is fairly large mostly to reduce the storage and service pressure 
+on metadata service. Since this Alluxio data cache keeps track of data and metadata locally, we significantly reduced the caching granularity, 
+termed as a ‘page’, for higher storage efficiency. This cache page size defaults to 1MB.
+
+### Cache location and hierarchy
+By default, Alluxio local cache stores data into the local filesystem. Each caching page is stored as a separate file under a directory structure as 
+follows:
+<br><br>
+*`<BASE_DIR>/LOCAL/1048576/<BUCKET>/<PAGE>`*
+<br><br>
+Here:<br>
+
+1. BASE_DIR is the root directory of the cache storage and is set by Presto configuration “cache.base-directory”.
+2. LOCAL means the cache storage type is LOCAL. Alluxio also supports rocksDB as the cache storage.
+3. 1048576: represents the 1MB block size.
+4. BUCKET represents a directory serving as buckets for various page files. They are created to make sure one single directory does not have too many 
+files which often leads to really bad performance. 
+5. PAGE represents the file named after the page id. In the presto the ID is the md5 hash of the filename. 
+
+### Thread Concurrency
+
+Each Presto worker keeps a set of threads, each executing different query tasks but sharing the same data cache. Thus this Alluxio data cache 
```

Comment:
Each Presto worker keeps a set of threads, each executing different query tasks, but sharing the same data cache.